PFW-1559 Add filament presence check at start of print

gudnimg opened the pull request:
Add filament presence check at start of print job. This PR aims to resolve issues where a user starts a print with no filament loaded. The idea is to use the M862 warning dialogs instead of relying on filament sensor runout. Relying on filament runout will require some fragile logic which may not be very reliable and risks adding new issues.

A new menu setting is added to configure how strict the check should be.

- None: No check is performed
- Warning **(default)**: The user is warned about missing filament and is prompted to continue with `Yes`/`No`. If `No` is selected, the print is aborted. If no user input is given (e.g. from host printing) then the warning will expire in 30 seconds and the printer assumes the `Yes` option was selected.
- Strict: If the filament is not detected when a print is started, it is immediately canceled with a message saying the filament is missing.

The check is currently performed on G-codes: M24 and M75. Starting a SD print and Host print respectively.

gudnimg committed:
Add filament presence check at start of print

A new menu setting is added to configure how strict the check should be.

- None: No check is performed
- Warning (default): The user is warned about missing and is prompted to continue with Yes/No. If No is selected, the print is aborted. If no user input is given (e.g. from host printing) then the warning will expire in 30 seconds and assume the Yes option was selected
- Strict: If the filament is not detected when a print is started, it is immediately canceled.

The check is currently performed on G-codes: M24 and M75. Starting a SD print and Host print respectively.

leptun commented on Firmware/eeprom.h:
```diff
@@ -245,7 +245,9 @@ static_assert(sizeof(Sheets) == EEPROM_SHEETS_SIZEOF, "Sizeof(Sheets) is not EEP
 | ^           | ^       | ^                                     | 02h 2        | ^                     | MMU cutter: __always__                            | ^            | ^
 | 0x0DAE 3502 | uint16  | EEPROM_UVLO_MESH_BED_LEVELING_FULL    | ???          | ff ffh 65535          | Saved MBL points                                  | Power Panic  | D3 Ax0dae C288
-| 0x0DAD 3501 | uint8   | _EEPROM_FREE_NR9_                     | ???          | ffh 255               | _Free EEPROM space_                               | _free space_ | D3 Ax0dad C1
+| 0x0DAD 3501 | uint8   | EEPROM_CHECK_FILAMENT                 | 01h 1        | ffh 255               | Check mode for filament is: __warn__              | LCD menu     | D3 Ax0dad C1
```
I don't think it's a good idea to reuse this. We can only reuse eeprom slots if we know for sure that the old value is ok or not initialized. This can only be done feasibly when a feature needs multiple eeprom bytes or when we don't care about the old value, such as in power panic.
So I suggest you consume a new eeprom byte for this.

gudnimg replied:
Done

leptun commented on Firmware/messages.cpp:
```diff
@@ -151,6 +151,8 @@ const char MSG_GCODE_NEWER_FIRMWARE_CONTINUE[] PROGMEM_I1 = ISTR("G-code sliced
 const char MSG_GCODE_DIFF_CONTINUE[] PROGMEM_I1 = ISTR("G-code sliced for a different level. Continue?"); ////MSG_GCODE_DIFF_CONTINUE c=20 r=3
 const char MSG_GCODE_DIFF_CANCELLED[] PROGMEM_I1 = ISTR("G-code sliced for a different level. Please re-slice the model again. Print cancelled."); ////MSG_GCODE_DIFF_CANCELLED c=20 r=8
+const char MSG_MISSING_FILAMENT_CONTINUE[] PROGMEM_I1 = ISTR("There is no filament loaded. Continue?"); ////MSG_MISSING_FILAMENT_CONTINUE c=20 r=3;
+const char MSG_MISSING_FILAMENT_CANCELLED[] PROGMEM_I1 = ISTR("There is no filament loaded. Print cancelled."); ////MSG_MISSING_FILAMENT_CANCELLED c=20 r=8
```
I'm not too fond of the duplicated messages. This is a good place to save some flash in the future.

gudnimg replied:
I think we should not tackle that in this PR as we discussed earlier today. But FYI the strings I have at the moment can be considered placeholders. I just typed something to create the initial PR. We can change the wording if we want.

gudnimg committed:
optimisation: deduplicate printer check toggle menus

This saves ~100B of flash

gudnimg committed:
optimisation: deduplicate more code

Change in memory:
Flash: -54 bytes
SRAM: 0 bytes

gudnimg committed:
Simplify filament_presence_check usage

Pull all checks needed into the function such that the caller
only needs to check the return value

Also added description for the filament_presence_check function

gudnimg committed:
Use a fresh EEPROM cell for the new value

gudnimg commented:
The following videos demonstrate the changes more visually 😃 


This video shows the default setting (Warning). I attempt to start a print with no filament loaded.
https://github.com/user-attachments/assets/334b5e05-1906-4891-9bea-38dbfcf5865a

This video shows the "Strict" setting. It will always cancel the print if filament is not loaded.
https://github.com/user-attachments/assets/d5a4fcf2-1716-4241-b88a-de5b2e829f15

leptun reviewed (approved):
Tested on MK3 (PAT9125). Works really nicely!

gudnimg commented:
Synced the PR with latest MK3 branch. Also got rid of the "fixup" commit to make the history a bit cleaner. (No code has been changed)

gudnimg commented:
Please do not merge this yet. I want to remove the mmu2.h include in utils.cpp. To make unit testing easier to begin with. 

To check if the MMU is enabled, I think it is ok to read the EEPROM value. The printer check shouldnt care if the MMU is communicating or not, it only cares if it is _enabled_ at all.

gudnimg committed:
Check if MMU is enabled by reading EEPROM

gudnimg commented:
Should be ready for review again

gudnimg committed:
fixup after rebase

3d-gussner committed:
Fix LCD output
Configuration->HW Setup->Checks showed `a`

MSG_FILAMENT `Filament` can only be 8 chars long and only ES and IT had `Filemanto` as translation.
But as ES the `o` in `Filamento` was overwritten by `[Estricto]` I removed the MSG_FILAMENT to be translated.
Saves few bytes.

3d-gussner committed:
Add missing translations
Used `trans -b :cs+de+es+fr+hr+hu+it+nl+no+pl+ro+sk+sv "<text>"` to translate
Looked at pre-translated text to reuse as much as possible.

3d-gussner reviewed (changes requested):
Please check https://github.com/gudnimg/Prusa-Firmware/pull/8

3d-gussner committed:
Update Romainian translations
Thanks to @petrubecheru and @Hauzman

3d-gussner committed:
Update Dutch translation
Thanks to @stelgenhof

3d-gussner committed:
Another Dutch translation fix

3d-gussner committed:
Update Hungarian translation
Thanks to @AttilaSVK

3d-gussner committed:
Update Czech and Italian translations
Thanks to @DRRacer and @wavexx

3d-gussner committed:
Update Slovak translation
Thanks to @ingbrzy @shatter136

3d-gussner committed:
Update Community translation list
- Croatian, Swedish and Norwegian are inactive since the last few translation review requests have not been done

gudnimg committed:
Merge pull request #8 from 3d-gussner/MK3_fix_PR4727

Fix LCD output and add translations
